use inference data in end of sampling report

OriolAbril opened the pull request:
Current code calls `az.ess` and `az.rhat` using trace objects instead of inference data objects. ArviZ therefore converts the trace to inferencedata object twice. This is generally not an issue and it is barely noticeable unless the number of observations is large, in which case the conversion (and consequent retrieving of pointwise log likelihood data) can be quite memory expensive. Below there is memory usage in an example (courtesy of @nitishp25) where the effect is quite noticeable:

![image](https://user-images.githubusercontent.com/23738400/79503494-8a48c880-8031-11ea-8a7d-5a4d000dc250.png)

This PR explicits the conversion to inferencedata so it only happens once, and if possible, skips retrieving the pointwise log likelihood data (which is not needed for ess nor rhat calculation).

OriolAbril commented on pymc3/backends/report.py:
```diff
@@ -107,7 +107,7 @@ def _run_convergence_checks(self, trace, model):
             return
 
-        from pymc3 import rhat, ess
+        from pymc3 import rhat, ess, _to_arviz
```
I have used this `to_arviz` approach for readability, but I can change it to `from arviz import from_pymc3` if you prefer it

michaelosthege replied:
Could `_to_arviz` be a place where to put storing of sampling metadata? (See https://github.com/arviz-devs/arviz/issues/1146)

OriolAbril replied:
I definitely think it has to be added to `from_pymc3` so also to `_to_arviz`, is that the question?

michaelosthege replied:
I didn't know `_to_arviz` existed, so to me it sounds like code duplication?

OriolAbril replied:
I aliased `from_pymc3` to `_to_arviz`, I was worried calling a function called `from_pymc3` inside pymc code would be confusing. I guess it depends on the general degree of familiarity with ArviZ

michaelosthege commented:
@OriolAbril do you think we could transition to returning `InferenceData` in the long term?

OriolAbril commented:
> @OriolAbril do you think we could transition to returning `InferenceData` in the long term?

Now that ArviZ is a dependency, it is feasible, however it would be quite a breaking change. It should probably be discussed in an issue or lab meeting. Another alternative would be to add an option to `pm.sampling` to return an inference data object :thinking:

AlexAndorra commented:
I like the idea of having an option to return an inference data object! Would make the modeling workflow easier and more elegant.

rpgoldman commented:
Returning an `InferenceData` is a nice idea, but it would radically change the posterior predictive sampling code, which is pretty rickety already.  To be honest, I'm so scarred by my experience working on `fast_sample_posterior_predictive` that I wouldn't like to go back in there again.

michaelosthege commented:
> 
> 
> Returning an `InferenceData` is a nice idea, but it would radically change the posterior predictive sampling code, which is pretty rickety already. To be honest, I'm so scarred by my experience working on `fast_sample_posterior_predictive` that I wouldn't like to go back in there again.

But we already have a converter to swallow `InferenceData.posterior` objects in sample_posterior_predictive - we did thata few weeks ago, remember?
https://github.com/pymc-devs/pymc3/blob/e7bc8326541e2c7aeee1f3741ff1877f716077a1/pymc3/sampling.py#L1561

rpgoldman commented:
Ah, right. I had forgotten. And it's already ported to `fast_sample_posterior_predictive`.  OK, never mind!

lucianopaz commented:
> > @OriolAbril do you think we could transition to returning InferenceData in the long term?
> 
> Now that ArviZ is a dependency, it is feasible, however it would be quite a breaking change. It should probably be discussed in an issue or lab meeting. Another alternative would be to add an option to pm.sampling to return an inference data object 🤔

I think that the steps to transition to an `InferenceData` interface should be:
1. Write a new trace backend that uses xarray datasets or whatever data container `arviz` uses under the hood. This way, arviz would only need to get a reference to the true trace instead of copying the contents into a new data structure.
2. Proceed as the guys in scikit-learn do: issue a `FutureWarning` in `pm.sample` saying that the default backend will change from `NDArray` to `XArray` (or whatever it is called in the end).
3. Leave the warning for the duration of at least one minor release before switching the default. By this I mean that, if the new backend is added in pymc3 v3.9, the default should change in v3.10.

The way in which the future warnings is dealt with usually is by setting the default argument value to something that indicates that the user did not explicitly provide anything as it's value (could be `None` or a string like `"unset"`). Then, if the argument is unset, the future warning is signaled and the argument's value is set to `NDArray` to preserve backwards compatibility.

nitishp25 commented:
Btw, this is how the memory usage looks after the changes done in this PR:

![pymc3_issue](https://user-images.githubusercontent.com/19911654/79636129-16263600-8193-11ea-9423-66ff0d26e867.png)

michaelosthege commented on pymc3/backends/report.py:
```diff
@@ -119,8 +119,12 @@ def _run_convergence_checks(self, trace, model):
-        self._rhat = rhat = rhat(trace, var_names=varnames)
+        try:
+            idata = _to_arviz(trace, log_likelihood=False)
+        except TypeError: # log_likelihood arg requires ArviZ >= 0.7.0
```
Maybe do this with
```
if az.__version__ >= '0.7.0':
    idata = az.from_pymc3(trace, log_likelihood=False)
else:
    idata = az.from_pymc3(trace)
```
(without the alias is less confusing IMHO)

rpgoldman replied:
Wouldn't it be just as easy to make arviz greater than 0.7 be an installation requirement?

Allowing arbitrary arviz versions is going to be very confusing going forward because of the tight coupling.

twiecki replied:
Agree with @rpgoldman, the way out is forward.

OriolAbril committed:
use from_pymc3(..., log_likelihood=False) and update requirements

OriolAbril commented:
Addressed the comments about using inference data to avoid memory usage peaks due to pointwise log likelihood. I would leave returning an inference data object as a result of `pm.sample` for another PR.

twiecki commented:
@OriolAbril Looks great -- thanks! Can you add a note to the release-notes under maintenance?

OriolAbril committed:
update release notes

twiecki commented:
Thanks @OriolAbril !
